pkg/core/mpt/trie.go

```diff
@@ -188,17 +220,37 @@ func (t *Trie) putIntoExtension(curr *ExtensionNode, path []byte, val Node) (Nod
 
 // putIntoHash puts val to trie if current node is a HashNode.
 // It returns Node if curr needs to be replaced and error if any.
-func (t *Trie) putIntoHash(curr *HashNode, path []byte, val Node) (Node, error) {
+func (t *Trie) putIntoHash(curr *HashNode, path []byte, val Node, restore bool) (Node, error) {
 	if curr.IsEmpty() {
+		if restore {
+			return nil, fmt.Errorf("%w: can't restore empty HashNode", ErrRestoreFailed)
+		}
 		hn := t.newSubTrie(path, val, true)
 		return hn, nil
 	}
 
 	result, err := t.getFromStore(curr.hash)
+	if restore && len(path) == 0 {
+		if err == nil {
+			// it's OK to have this node in the storage (i.e. if this part of MPT is untouched after
+			// previous changes), because we don't remove old MPT nodes while syncing MPT state via P2P,
+			// so just add it to local MPT.
```

Comment:
Do we need to get it from storage in the first place (before this `if`)? It seems like the only thing we achieve with this `getFromStore` here is checking if our storage works properly.

Reply:
You're right, fixed.